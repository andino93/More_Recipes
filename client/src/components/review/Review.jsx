import React, { Component } from 'react';
import PropTypes from 'react-proptypes';
import { connect } from 'react-redux';
import classnames from 'classnames';
import swal from 'sweetalert2';
import { bindActionCreators } from 'redux';
import ReviewList from './ReviewList.jsx';
import { postReview } from '../../actions/recipe/recipeActions';
import reviewValidate from "../../utils/reviewValidate";
import { recipeResponseType, decodeToken } from '../../utils/helpers';

/**
 *
 * @export
 * 
 * @class Review
 * 
 * @extends {React.Component}
 */
export class Review extends Component {
  /**
   * Creates an instance of ReviewComponent.
   * 
   * @param {any} props
   * 
   * @memberof AddReview
   */
  constructor(props) {
    super(props);
    this.state = {
      content: '',
      errors: {}
    };
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  /**
   *
   * @param  {object} nextProps
   * 
   * @memberof RecipeDetail
   * 
   * @returns {XML} XML/JSX
   * 
   */
  componentWillReceiveProps(nextProps) {
    const { responseType } = nextProps.reviewState;
    if (responseType === recipeResponseType.POST_REVIEW_SUCCESS) {
      const reviewInfo = {
        content: this.state.content,
        createdAt: new Date(),
        users: {
          username: decodeToken(localStorage.token).username
        }
      };
      this.props.reviews.unshift(reviewInfo);
      this.setState({ content: '' });
      swal({
        position: 'top-end',
        title: 'Awesome !',
        text: 'Your review was successfully posted !',
        type: 'success',
        showConfirmButton: false,
        timer: 1500
      });
    }
  }

  /**
   * Handle change
   *
   * @param {event} event
   * 
   * @return {event} event
   * 
   */
  handleChange(event) {
    this.setState({ content: event.target.value });
  }

  /**
   * Handles review creation
   * 
   * @param { object } event
   * 
   * @method handleSubmit
   * 
   * @return {void}
   */
  handleSubmit(event) {
    event.preventDefault();
    if (this.checkValidity()) {
      this.setState({ errors: {} });
      const { recipeId } = this.props.match.params;
      this.props.postReview(recipeId, this.state.content);
    }
  }

  /** Check validity
   * 
   * @method checkValidity
   * 
   * @return {object} isValid
   */
  checkValidity() {
    const { errors, isValid } = reviewValidate(this.state);
    if (!isValid) {
      this.setState({ errors });
    }
    return isValid;
  }
  /**
   *
   * @memberof AddReview
   * 
   * @returns {XML} XML/JSX
   * 
   */
  render() {
    const { errors } = this.state;
    return (
      <div>
        <form onSubmit={this.handleSubmit}>
          <div className="form-group">
            <h5 className="text-muted">
              <label htmlFor="review" className="row justify-content-center">
                Please post a review
              </label>
            </h5>
            <textarea className ={
              classnames('form-control',
                {
                  'is-invalid': errors.content ? !!errors.content : false
                })
            }
            rows="3"
            onChange={this.handleChange}
            value={this.state.content}
            placeholder="Kindly enter your review"
            />
            { errors.content ?
              <span className="invalid-feedback">{ errors.content }
              </span> : null
            }
          </div>
          <button type="submit"
            className="btn btn-primary mb-2">
            Submit Review
          </button>
        </form>

        <h5 className="p-5 text-center">Reviews</h5>
        { this.props.reviews.map((review, i) => (
          <ReviewList
            key={i}
            review = {review}/>
        )) }
      </div>
    );
  }
}
Review.propTypes = {
  reviewState: PropTypes.object.isRequired,
  postReview: PropTypes.func.isRequired,
  reviews: PropTypes.array,
  match: PropTypes.object
};

const mapStateToProps = state => ({
  reviewState: state.recipeReducer
});

const mapDispatchToProps = dispatch =>
  bindActionCreators({ postReview }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Review);