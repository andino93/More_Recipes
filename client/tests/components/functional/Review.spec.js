import React from 'react';
import expect from 'expect';
import Adapter from 'enzyme-adapter-react-16';
import { configure } from 'enzyme';
import mockStore from '../../__mocks__/mockStore';
import mockData from '../../__mocks__/mockData';
import mountWrap from '../../__mocks__/utils';
import localStorageMock from '../../__mocks__/localStorageMock';
import Review from '../../../src/components/review/Review.jsx';

configure({ adapter: new Adapter() });

// noinspection JSAnnotator
window.localStorage = localStorageMock;
window.localStorage.token = mockData.token;

describe('<Review>', () => {
  const store = mockStore({ recipeReducer: { reviewState: {} } });
  const connectedReview = mountWrap(Review, store, {
    reviews: [],
    reviewState: {},
    match: {
      params: {
        recipeId: 1
      }
    }
  });
  it('should render the Review Component', (done) => {
    expect(connectedReview).toMatchSnapshot();
    done();
  });
  it('should check for componentWillReceiveProps', () => {
    connectedReview.find('Review').instance().componentWillReceiveProps({
      reviewState: {
        responseType: 'POST_REVIEW_SUCCESS',
      },
    });
  });
  it('should simulate handleChange', (done) => {
    connectedReview.find('textarea').at(0).simulate('change');
    done();
  });
  it('should simulate handleClick', (done) => {
    connectedReview.find('form').at(0).simulate('submit', {
      preventDefault: () => {}
    });
    connectedReview.find('Review').instance().setState({
      content: 'Amazing'
    });
    connectedReview.find('form').at(0).simulate('submit', {
      preventDefault: () => {}
    });
    done();
  });
});