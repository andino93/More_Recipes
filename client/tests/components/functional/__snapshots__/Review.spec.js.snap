// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Review> should render the Review Component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "blank",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(Review)
        match={
          Object {
            "params": Object {
              "recipeId": 1,
            },
          }
        }
        reviewState={Object {}}
        reviews={Array []}
      >
        <Review
          match={
            Object {
              "params": Object {
                "recipeId": 1,
              },
            }
          }
          postReview={[Function]}
          reviewState={
            Object {
              "reviewState": Object {},
            }
          }
          reviews={Array []}
        >
          <div>
            <form
              onSubmit={[Function]}
            >
              <div
                className="form-group"
              >
                <h5
                  className="text-muted"
                >
                  <label
                    className="row justify-content-center"
                    htmlFor="review"
                  >
                    Please post a review
                  </label>
                </h5>
                <textarea
                  className="form-control"
                  onChange={[Function]}
                  placeholder="Kindly enter your review"
                  rows="3"
                  value=""
                />
              </div>
              <button
                className="btn btn-primary mb-2"
                type="submit"
              >
                Submit Review
              </button>
            </form>
            <h5
              className="p-5 text-center"
            >
              Reviews
            </h5>
          </div>
        </Review>
      </Connect(Review)>
    </Router>
  </BrowserRouter>
</Provider>
`;
